doc/design/figures/xgboost_feature_column.md

```diff
@@ -0,0 +1,131 @@
+| SEQ_CATEGORY_ID  | tf.feature_column.sequence_categorical_column_with_identity | Sequence data version of CATEGORY_ID                                 |
+| CROSS            | tf.feature_column.crossed_column                            | Combine multiple category features using hash algorithm              |
+| INDICATOR        | tf.feature_column.indicator_column                          | Transform category id to multi-hot representation                    |
+| EMBEDDING        | tf.feature_column.embedding_column                          | Transform category id to embedding representation                    |
+
+## Feature column design in XGBoost models
+
+The training process of XGBoost model inside SQLFlow are as follows:
+
+- Step 1: Read data from database. Call the `db.db_generator()` method to return a Python generator which yields each row in database. 
+- Step 2: Dump SVM file. Call the `dump_dmatrix()` method to write the raw data into an SVM file. This file is ready to be loaded as XGBoost DMatrix.
+- Step 3: Training. Load the dumped file to be XGBoost DMatrix and start to train.
+
+Between step 1 and 2, we can do some pre-processing to the raw data from database. In Python side, we can implement the feature column interfaces like:
+
+```Python
+class BaseFeatureColumnTransformer(object):
+    def __call__(self, inputs):
+        raise NotImplementedError()
+
+class NumericColumnTransformer(BaseFeatureColumnTransformer):
+    # `column_idx` is the feature index inside `SELECT` statement
+    def __init__(self, column_idx):
+        self.column_idx = column_idx
+    
+    # `inputs` are all raw column data
+    # NumericColumnTransformer would only take the column indicated by `column_idx`
+    def __call__(self, inputs):
+        return inputs[self.column_idx]
+    
+# CategoryColumnTransformer is the base class of all category columns
+# This base class is design to do some check. For example, `INDICATOR`
+# would only accept category column as its input.
+class CategoryColumnTransformer(BaseFeatureColumnTransformer): pass
+        
+class BucketizedColumnTransformer(CategoryColumnTransformer):
+    def __init__(self, column_idx, boundaries, default_value=None):
+        self.column_idx = column_idx
+        self.boundaries = boundaries
+        self.default_value = default_value
+        
+    def __call__(self, inputs):
+        input = inputs[self.column_idx]
+        if input < boundaries[0]:
+            return 0
+            
+        for idx, b in enumarate(boudaries):
+            if input >= b
+                return idx
+        return len(boundaries)
+
+# CrossedColumnTransformer is designed as a decorator of CategoryColumnTransformer
+# It accepts several CategoryColumnTransformers, and crosses their transformed
+# results using hash algorithm.
+class CrossedColumnTransformer(CategoryColumnTransformer):
```

Comment:
Cross columns should not be a subclass of `CategoryColumnTransformer` since it may not be a categorical column.

Reply:
Thanks. I have misunderstond `CrossColumn`. Fixed.